src/get.rs

```diff
@@ -129,14 +134,171 @@ impl AsyncRead for DataStream {
     }
 }
 
+/// Gets a collection and all its blobs using a [`Ticket`].
+pub async fn run_ticket<A, B, C, FutA, FutB, FutC>(
+    ticket: Ticket,
+    keylog: bool,
+    max_concurrent: u8,
+    on_connected: A,
+    on_collection: B,
+    on_blob: C,
+) -> Result<Stats>
+where
+    A: FnOnce() -> FutA,
+    FutA: Future<Output = Result<()>>,
+    B: FnOnce(&Collection) -> FutB,
+    FutB: Future<Output = Result<()>>,
+    C: FnMut(Hash, DataStream, String) -> FutC,
+    FutC: Future<Output = Result<DataStream>>,
+{
+    let start = Instant::now();
+    let connection = dial_ticket(&ticket, keylog, max_concurrent.into()).await?;
+    run_connection(
+        connection,
+        ticket.hash,
+        ticket.token,
+        start,
+        on_connected,
+        on_collection,
+        on_blob,
+    )
+    .await
+}
+
+async fn dial_ticket(
+    ticket: &Ticket,
+    keylog: bool,
+    max_concurrent: usize,
+) -> Result<quinn::Connection> {
+    // Sort the interfaces to make sure local ones are at the front of the list.
+    let interfaces = default_net::get_interfaces();
+    let mut addrs: Vec<SocketAddr> = ticket
+        .addrs
+        .iter()
+        .filter(|addr| is_same_subnet(addr, &interfaces))
+        .copied()
+        .collect();
+    let other_addrs: Vec<SocketAddr> = ticket
+        .addrs
+        .iter()
+        .filter(|addr| !addrs.contains(addr))
+        .copied()
+        .collect();
```

Comment:
you could use https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition instead of iterating twice